File: java/src/dsclient/scheduling_policies/CostEffective.java
package dsclient.scheduling_policies;

import java.io.IOException;
import java.util.Comparator;
import java.util.List;

import dsclient.JobDispatchPolicy;
import dsclient.JobScheduler;
import dsclient.enums.ServerState;
import dsclient.models.JobSubmission;
import dsclient.models.Resource;

public class CostEffective implements JobDispatchPolicy {
    public void dispatch(JobScheduler scheduler) throws IOException {
        while (true) {
            String received = scheduler.inquire("REDY");
            if (received.equals("NONE")) {
                break;
            }

            JobSubmission job = JobSubmission.fromReceivedLine(received);

            List<Resource> resources = scheduler.fetchCapableResources(job);

            Resource best = resources.stream().max(
                Comparator.<Resource>comparingInt(
                    res -> -(res.coreCount - job.coreCount)
                ).thenComparingInt(
                    res -> (res.state == ServerState.IDLE) ? 1 : 0
                ).thenComparingInt(res -> -res.availableTime)
            ).get();

            scheduler.schedule(job, best);
        }
    }
}

File: python/ds_client/scheduling_policies/cost_effective.py

from __future__ import annotations
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from ..job_scheduler import JobScheduler

from ..job_dispatch_policy import JobDispatchPolicy
from ..models import JobSubmission
from ..enums import ServerState

class CostEffective(JobDispatchPolicy):
    def dispatch(self, scheduler: JobScheduler) -> None:
        while True:
            received = scheduler.inquire('REDY')
            if received == 'NONE':
                break

            job = JobSubmission.from_received_line(received)

            resources = scheduler.fetch_capable_resources(job)

            best = max(
                resources,
                key=lambda res: (
                    -(res.core_count - job.core_count),
                    res.state == ServerState.IDLE,
                    -res.available_time,
                ),
            )

            scheduler.schedule(job, best)
